File: Sources/PetCompanion/Preferences/AppPreferences.swift
import Foundation

final class AppPreferences {
    private enum Keys {
        static let isPaused = "isPaused"
        static let petSpeed = "petSpeed"
        static let petScale = "petScale"
        static let showInFullScreen = "showInFullScreen"
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        registerDefaults()
    }

    var isPaused: Bool {
        get { defaults.bool(forKey: Keys.isPaused) }
        set { defaults.set(newValue, forKey: Keys.isPaused) }
    }

    var petSpeed: Double {
        get { defaults.double(forKey: Keys.petSpeed) }
        set { defaults.set(newValue, forKey: Keys.petSpeed) }
    }

    var petScale: Double {
        get { defaults.double(forKey: Keys.petScale) }
        set { defaults.set(newValue, forKey: Keys.petScale) }
    }

    var showInFullScreen: Bool {
        get { defaults.bool(forKey: Keys.showInFullScreen) }
        set { defaults.set(newValue, forKey: Keys.showInFullScreen) }
    }

    private func registerDefaults() {
        defaults.register(defaults: [
            Keys.isPaused: false,
            Keys.petSpeed: 82.0,
            Keys.petScale: 1.0,
            Keys.showInFullScreen: false
        ])
    }
}

File: Windows/PetCompanion.Windows/Pet/AppPreferences.cs
using System;
using System.IO;
using System.Text.Json;

namespace PetCompanion.Windows.Pet;

public sealed class AppPreferences
{
    private static readonly string DirectoryPath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "PetCompanion");
    private static readonly string FilePath = Path.Combine(DirectoryPath, "preferences.json");

    public bool IsPaused { get; set; }
    public double PetSpeed { get; set; } = 82;
    public double PetScale { get; set; } = 1;
    public bool ShowInFullScreen { get; set; }
    public bool LaunchAtLogin { get; set; }

    public AppPreferences()
    {
        Load();
    }

    public void Save()
    {
        Directory.CreateDirectory(DirectoryPath);
        var data = new PreferenceData(IsPaused, PetSpeed, PetScale, ShowInFullScreen, LaunchAtLogin);
        var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(FilePath, json);
    }

    private void Load()
    {
        if (!File.Exists(FilePath))
        {
            return;
        }

        try
        {
            var loaded = JsonSerializer.Deserialize<PreferenceData>(File.ReadAllText(FilePath));
            if (loaded is null)
            {
                return;
            }

            IsPaused = loaded.IsPaused;
            PetSpeed = loaded.PetSpeed;
            PetScale = loaded.PetScale;
            ShowInFullScreen = loaded.ShowInFullScreen;
            LaunchAtLogin = loaded.LaunchAtLogin;
        }
        catch
        {
            // Keep defaults when the user config is unreadable.
        }
    }

    private sealed record PreferenceData(
        bool IsPaused,
        double PetSpeed,
        double PetScale,
        bool ShowInFullScreen,
        bool LaunchAtLogin);
}
